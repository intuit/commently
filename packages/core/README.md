# commently-core

Commently Core is the underlying API package wrapped by the CLI

## Usage

To use, you can simply import it into your project like so. Remember that you will still need to provide a `GITHUB_TOKEN` or `GH_TOKEN`, as well as a `GITHUB_URL` (default to public github) in your environment to be able to comment.

```javascript
import Commently from 'commently/core';

const commently = new Commently({
  pr: '1234',
  owner: 'GithubOrg',
  repo: 'RepoName',
  title: 'The Title of Your PR Comment',
  key: 'unique-id'
});

commently
  .autoComment('The body of your PR')
  .then(response => {
    console.log(symbols.success, `Successfully commented on ...}`);
    console.log(response.data.html_url);
  })
  .catch(err => {
    console.log(symbols.error, 'Oops! Something went wrong...');
    console.log(err);
  });
```

## Gitlab

We also now export an API compliant gitlab module. It needs a `GITLAB_TOKEN` to exist in your environment. It also needs a commit sha to fine your Merge Request.

```javascript
import { Gitlab } from 'commently/core';

const commently = new Gitlab({
  commit: 'commit-sha',
  title: 'The Title of Your MR Comment'
});

commently
  .autoComment('The body of your MR')
  .then(response => {
    console.log(symbols.success, `Successfully commented on ...}`);
  })
  .catch(err => {
    console.log(symbols.error, 'Oops! Something went wrong...');
    console.log(err);
  });
```
