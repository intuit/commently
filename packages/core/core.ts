import * as Octokit from '@octokit/rest';
import * as debug from 'debug';
import * as envCi from 'env-ci';

interface CommentlyArgs {
  pr?: number;
  owner?: string;
  repo?: string;
  title?: string;
}

interface EnvCi {
  pr?: number;
  slug?: string;
}

interface User {
  id: number;
}

export default class Commently {
  public readonly header: string;

  private user?: User;
  private readonly octokit: Octokit;
  private readonly owner: string;
  private readonly repo: string;
  private readonly title: string;
  private readonly issueId: number;
  private readonly footer: string;
  private readonly delim: string;
  private readonly debug: debug.IDebugger;

  constructor(args: CommentlyArgs) {
    this.debug = debug('commently');

    const { pr, slug = '' } = envCi() as EnvCi;
    const [owner, repo] = slug.split('/');
    const prNumber = args.pr || pr;

    if (!prNumber) {
      throw new Error(
        "A PR number wasn't provided as an argument (--pr) or detected in the CI environment."
      );
    }

    this.title = args.title || 'commently';
    this.owner = args.owner || owner;
    this.repo = args.repo || repo;

    if (!this.owner) {
      throw new Error(
        "A owner wasn't provided as an argument (--owner) or detected in the CI environment."
      );
    }

    if (!this.repo) {
      throw new Error(
        "A repo wasn't provided as an argument (--repo) or detected in the CI environment."
      );
    }

    this.issueId = prNumber;
    this.header = `<!-- \n ${this.title}-id: ${this.issueId} \n -->\n${
      this.title
    }\n`;
    this.footer = `Courtesy of your **[${
      this.title
    }](https://github.com/intuit/commently)** bot :package::rocket:`;
    this.delim = `<!-- ${this.title}-section -->\n\n`;

    this.debug('Initialized: owner=%s repo=%s', this.owner, this.repo);

    const startUpArgs: Octokit.Options = {
      baseUrl: process.env.GITHUB_URL || 'https://api.github.com'
    };
    const token = process.env.GH_TOKEN || process.env.GITHUB_TOKEN;

    if (token) {
      startUpArgs.auth = `token ${token}`;
    }

    this.octokit = new Octokit(startUpArgs);
  }

  async autoComment(body: string, append = true) {
    this.debug('Auto Commenting: issueId=%s append?=%s', this.issueId, append);

    const keyedComment = await this.getKeyedComment();

    if (!keyedComment) {
      return this.createKeyedComment(body);
    }

    return this.editKeyedComment(keyedComment, body, append);
  }

  private async editKeyedComment(
    comment: Octokit.IssuesListCommentsResponseItem,
    body: string,
    append = true
  ) {
    this.debug('Editing existing Keyed Comment. id=%s', comment.id);
    const historyDelim = `<!-- ${this.title}-history -->`;

    if (append) {
      const parts = comment.body.split(this.delim);
      const header = parts.shift();
      const last = (parts.shift() || '')
        .replace('- ', '')
        .replace(/`/g, '')
        .replace('@', '<span>@</span>');
      const footer = parts.pop();
      let history = parts.shift();

      if (history) {
        const historyParts = history.split(historyDelim);
        const historyHeader = historyParts.shift();
        const historyFooter = historyParts.pop();
        // Only keep 10 items in history
        const historySubParts = (historyParts.pop() || '').split('<hr>');

        if (historySubParts.length > 10) {
          historySubParts.splice(0, historySubParts.length - 10);
        }

        historyParts.push(historySubParts.join('<hr>'));
        historyParts.push(`<hr><p>${last.split('\n').join('<br>')}</p>`);
        history = [historyHeader, historyParts.join('\n'), historyFooter].join(
          `${historyDelim}`
        );
      } else {
        history = `<details><summary>History</summary><div>\n${historyDelim}\n<hr><p>${last
          .split('\n')
          .join('<br>')}</p>${historyDelim}\n</div></details>\n</br>`;
      }

      const newComment = [header, `${body}\n`, history, footer].filter(
        (part): part is string => typeof part === 'string'
      );
      parts.push(...newComment);

      return this.editComment(comment.id, parts.join(this.delim));
    }

    return this.editComment(
      comment.id,
      `${this.header}${this.delim}${body}\n${this.delim}${this.footer}`
    );
  }

  private async editComment(id: number, body: string) {
    return this.octokit.issues.updateComment({
      body,
      // eslint-disable-next-line
      comment_id: id,
      owner: this.owner,
      repo: this.repo
    });
  }

  private async getKeyedComment() {
    const allComments = await this.getComments();
    const user = await this.getUser();
    const userComments = allComments.filter(
      comment => comment.user.id === user.id
    );
    const keyedComments = userComments.filter(
      comment => comment.body.indexOf(this.header) !== -1
    );

    return keyedComments[0];
  }

  private async getComments(
    options: Partial<Octokit.IssuesListCommentsParams> = {}
  ) {
    const response = await this.octokit.issues.listComments({
      number: this.issueId,
      owner: this.owner,
      repo: this.repo,
      ...options
    });

    if (!response || !response.data) {
      throw new Error(`Could not get comments for issue: ${this.issueId}`);
    }

    return response.data;
  }

  private async getUser() {
    if (this.user) {
      return this.user;
    }

    const response = await this.octokit.users.getAuthenticated({});

    if (response && response.data) {
      this.user = response.data as User;
      return this.user;
    }

    throw new Error('Could not retrieve Github user');
  }

  private async createComment(body: string) {
    return this.octokit.issues.createComment({
      body,
      number: this.issueId,
      owner: this.owner,
      repo: this.repo
    });
  }

  private async createKeyedComment(body: string) {
    this.debug('Creating new Keyed Comment');

    return this.createComment(
      `${this.header}${this.delim}${body}\n${this.delim}${this.footer}`
    );
  }
}
