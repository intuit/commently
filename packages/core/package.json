{
  "name": "commently",
  "version": "0.0.2",
  "description": "Easily comment on github issues and PRs",
  "main": "dist/core.js",
  "scripts": {
    "test": "jest",
    "lint": "eslint **/*.ts",
    "build": "tsc"
  },
  "dependencies": {
    "@octokit/rest": "^16.13.3",
    "debug": "^4.1.1",
    "env-ci": "^3.2.0"
  },
  "devDependencies": {
    "@types/debug": "0.0.31",
    "@types/env-ci": "3.1.0",
    "@types/node": "10.12.27"
  }
}
