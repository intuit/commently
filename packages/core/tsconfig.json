{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist"
  },
  "include": ["src/core.ts"]
}
