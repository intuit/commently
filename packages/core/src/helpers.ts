import {Comment, Github, Gitlab} from './core'
import * as Octokit from '@octokit/rest';

type TComment = Comment | Octokit.IssuesListCommentsResponseItem
type TThis = Github | Gitlab

export function buildKeyedComment (
  this: TThis,
  comment: TComment,
  body: string,
  append = true,
) {
  this.debug('Editing existing Keyed Comment. id=%s', comment.id);
  const historyDelim = `<!-- ${this.key}-history -->`;

  if (append) {
    const parts = comment.body.split(this.delim);
    const header = parts.shift();
    const last = (parts.shift() || '')
      .replace('- ', '')
      .replace(/`/g, '')
      .replace('@', '<span>@</span>');
    const footer = parts.pop();
    let history = parts.shift();

    if (history) {
      const historyParts = history.split(historyDelim);
      const historyHeader = historyParts.shift();
      const historyFooter = historyParts.pop();
      // Only keep 10 items in history
      const historySubParts = (historyParts.pop() || '').split('<hr>');

      if (historySubParts.length > 10) {
        historySubParts.splice(0, historySubParts.length - 10);
      }

      historyParts.push(historySubParts.join('<hr>'));
      historyParts.push(`<hr><p>${last.split('\n').join('<br>')}</p>`);
      history = [historyHeader, historyParts.join('\n'), historyFooter].join(
        `${historyDelim}`
      );
    } else {
      history = `<details><summary>History</summary><div>\n${historyDelim}\n<hr><p>${last
        .split('\n')
        .join('<br>')}</p>${historyDelim}\n</div></details>\n</br>`;
    }

    const newComment = [
      header,
      `${body}\n`,
      this.useHistory && history,
      footer
    ].filter((part): part is string => typeof part === 'string');
    parts.push(...newComment);

    return {
      id: comment.id,
      body: parts.join(this.delim)
    };
  }

  return {
    id: comment.id,
    body: [this.header, body+'\n', this.footer].join(this.delim),
  }
}