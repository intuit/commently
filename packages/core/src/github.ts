/* eslint-disable camelcase */

import * as Octokit from '@octokit/rest';
import * as debug from 'debug';
import * as envCi from 'env-ci';
import { buildKeyedComment } from './helpers';

interface CommentlyArgs {
  /** The PR to comment on. Is detected in CI environments */
  pr?: number;
  /** The owner fo the repo to comment on. Is detected in CI environments */
  owner?: string;
  /** The repo to comment on. Is detected in CI environments */
  repo?: string;
  /** The title at the top of the comment */
  title?: string;
  /** The unique key to identify the comment by, not shown to end users */
  key?: string;
  /** Key a history of the comments in the comment created by this library */
  useHistory?: boolean;
}

interface User {
  id: number;
}

/** Create a "commenter" that can comment on a pull request */
export class Github {
  public readonly header: string;
  public readonly useHistory: boolean;
  public readonly key: string;
  public readonly footer: string;
  public readonly delim: string;
  public readonly debug: debug.IDebugger;

  private user?: User;
  private readonly octokit: Octokit;
  private readonly owner: string;
  private readonly repo: string;
  private readonly title: string;
  private readonly issueId: number;

  constructor(args: CommentlyArgs) {
    this.debug = debug('commently');

    const env = envCi();
    const slug = ('slug' in env && env.slug) || '';
    const [owner, repo] = slug.split('/');
    const prNumber = args.pr || ('pr' in env && Number(env.pr));

    if (!prNumber) {
      throw new Error(
        "A PR number wasn't provided as an argument (--pr) or detected in the CI environment."
      );
    }

    this.title = args.title || '';
    this.key = args.key || 'commently';
    this.owner = args.owner || owner;
    this.repo = args.repo || repo;
    this.useHistory = 'useHistory' in args ? Boolean(args.useHistory) : true;

    if (!this.owner) {
      throw new Error(
        "A owner wasn't provided as an argument (--owner) or detected in the CI environment."
      );
    }

    if (!this.repo) {
      throw new Error(
        "A repo wasn't provided as an argument (--repo) or detected in the CI environment."
      );
    }

    this.issueId = prNumber;
    this.header = `<!-- \n ${this.key}-id: ${this.issueId} \n -->\n${this.title}\n`;
    this.footer = `Courtesy of your **[${this.key}](https://github.com/intuit/commently)** bot :package::rocket:`;
    this.delim = `<!-- ${this.key}-section -->\n\n`;

    this.debug('Initialized: owner=%s repo=%s', this.owner, this.repo);

    const startUpArgs: Octokit.Options = {
      baseUrl: process.env.GITHUB_URL || 'https://api.github.com'
    };
    const token = process.env.GH_TOKEN || process.env.GITHUB_TOKEN;

    if (token) {
      startUpArgs.auth = `token ${token}`;
    }

    this.octokit = new Octokit(startUpArgs);
  }

  /**
   * Comment the body param on the pull requests
   *
   * @param {string} body - the comment to post
   * @param {boolean} append - whether to append the comment to the last comment
   */
  async autoComment(body: string, append = true) {
    this.debug('Auto Commenting: issueId=%s append?=%s', this.issueId, append);

    const keyedComment = await this.getKeyedComment();

    if (!keyedComment) {
      return this.createKeyedComment(body);
    }

    return this.editKeyedComment(keyedComment, body, append);
  }

  private async editKeyedComment(
    comment: Octokit.IssuesListCommentsResponseItem,
    bodyIn: string,
    append = true
  ) {
    const {id, body} = buildKeyedComment.bind(this)(comment, bodyIn, append)
    return this.editComment(
      id,
      body
    );
  }

  private async editComment(id: number, body: string) {
    return this.octokit.issues.updateComment({
      body,
      // eslint-disable-next-line
      comment_id: id,
      owner: this.owner,
      repo: this.repo
    });
  }

  private async getKeyedComment() {
    const allComments = await this.getComments();
    const user = await this.getUser();
    const userComments = allComments.filter(
      comment => comment.user.id === user.id
    );
    const keyedComments = userComments.filter(
      comment => comment.body.indexOf(this.header) !== -1
    );

    return keyedComments[0];
  }

  private async getComments(
    options: Partial<Octokit.IssuesListCommentsParams> = {}
  ) {
    const response = await this.octokit.issues.listComments({
      issue_number: this.issueId, // eslint-disable-line @typescript-eslint/camelcase
      owner: this.owner,
      repo: this.repo,
      ...options
    });

    if (!response || !response.data) {
      throw new Error(`Could not get comments for issue: ${this.issueId}`);
    }

    return response.data;
  }

  private async getUser() {
    if (this.user) {
      return this.user;
    }

    const response = await this.octokit.users.getAuthenticated({});

    if (response && response.data) {
      this.user = response.data as User;
      return this.user;
    }

    throw new Error('Could not retrieve Github user');
  }

  private async createComment(body: string) {
    return this.octokit.issues.createComment({
      body,
      issue_number: this.issueId, // eslint-disable-line @typescript-eslint/camelcase
      owner: this.owner,
      repo: this.repo
    });
  }

  private async createKeyedComment(body: string) {
    this.debug('Creating new Keyed Comment');

    return this.createComment(
      `${this.header}${this.delim}${body}\n${this.delim}${this.footer}`
    );
  }
}
