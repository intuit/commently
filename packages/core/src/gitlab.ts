/* eslint-disable camelcase, @typescript-eslint/camelcase */

import * as GitlabApi from 'gitlab';
import * as debug from 'debug';
import * as envCi from 'env-ci';
import { buildKeyedComment } from './helpers';

interface CommentlyArgs {
  /** The MR sha to comment on. Is detected in CI environments */
  commit?: string;
  /** The title at the top of the comment */
  title?: string;
  /** The unique key to identify the comment by, not shown to end users */
  key?: string;
  /** Key a history of the comments in the comment created by this library */
  useHistory?: boolean;
  /** The gitlab token if not in your environment */
  token?: string;
}

interface User {
  id: number;
}

interface MRInfo {
  project_id: number;
  iid: number;
  sha: string;
}

export interface Comment {
  author: {
    id: number;
  };
  body: string;
  sha: string;
  id: number;
}

/** Create a "commenter" that can comment on a pull request */
export class Gitlab {
  public readonly header: string;
  public readonly useHistory: boolean;
  public readonly key: string;
  public readonly footer: string;
  public readonly delim: string;
  public readonly debug: debug.IDebugger;

  private user?: User;
  private project_id?: number;
  private iid?: number;
  private readonly commitSha: string;
  private readonly gitlabToken: string;
  private readonly title: string;
  private readonly gitlab: GitlabApi.Gitlab;

  constructor(args: CommentlyArgs) {
    this.debug = debug('commently-gitlab');

    const env = envCi();
    const commitSha = args.commit || ('commit' in env && env.commit);
    const gitlabToken = process.env.GITLAB_TOKEN || args.token
    const gitlabUrl = process.env.GITLAB_URL || 'https://gitlab.com'
    const certsIgnored = process.env.NODE_TLS_REJECT_UNAUTHORIZED === "0"
    if (!commitSha) {
      throw new Error(
        "A Commit SHA wasn't provided as an argument (--commit) or detected in the CI environment.",
      );
    }

    if (!gitlabToken) {
      throw new Error(
        "A gitlab token wasn't provided as an argument (--token) or detected in your environment.",
      );
    }

    this.gitlabToken = gitlabToken
    this.commitSha = commitSha
    this.title = args.title || '';
    this.key = args.key || 'commently';
    this.useHistory = 'useHistory' in args ? Boolean(args.useHistory) : true;

    this.header = `<!-- \n ${this.key}-id: ${this.commitSha} \n -->\n${this.title}\n`;
    this.footer = `Courtesy of your **[${this.key}](https://github.com/intuit/commently)** bot :package: :rocket:`;
    this.delim = `<!-- ${this.key}-section -->\n\n`;

    this.debug('Initialized: commit=%s', this.commitSha);

    this.gitlab = new GitlabApi.Gitlab({
      token: this.gitlabToken,
      host: gitlabUrl,
      rejectUnauthorized: !certsIgnored
    });
  }

  /**
   * Comment the body param on the pull requests
   *
   * @param {string} body - the comment to post
   * @param {boolean} append - whether to append the comment to the last comment
   */
  async autoComment(body: string, append = true) {
    this.debug('Auto Commenting: commitSha=%s append?=%s', this.commitSha, append);

    const keyedComment = await this.getKeyedComment();

    if (!keyedComment) {
      return this.createKeyedComment(body);
    }

    return this.editKeyedComment(keyedComment, body, append);
  }

  private async editKeyedComment(
    comment: Comment,
    bodyIn: string,
    append = true
  ) {
    const {id, body} = buildKeyedComment.bind(this)(comment, bodyIn, append)
    return this.editComment(
      id,
      body
    );
  }

  private async editComment(id: number, body: string) {
    return this.gitlab.MergeRequestNotes.edit(
      this.project_id as number, this.iid as number, id, body,
    );
  }

  private async getKeyedComment() {
    const allComments = await this.getComments();
    const user = await this.getUser() as User;
    return allComments.find(
      (comment) => {
        const hasHeader = comment.body && comment.body.indexOf(this.header) !== -1
        const weWrote = comment.author && comment.author.id === user.id
        return Boolean(hasHeader) && Boolean(weWrote)
      },
    );
  }

  private async getMr() {
    const currentMergeRequests = await this.gitlab.MergeRequests.all({}) as MRInfo[]
    const shaMergeRequest = currentMergeRequests && currentMergeRequests.find(info => info.sha === this.commitSha)
    const {project_id, iid} = shaMergeRequest as MRInfo
    if (!project_id || !iid) {
      throw new Error(
        'Could not find any merge requests at all. Are you sure you have your token and sha correctly setup?',
      )
    }

    this.project_id = project_id
    this.iid = iid
  }

  private async getComments() {
    if (!this.project_id || !this.iid) {
      await this.getMr()
    }

    const currentMergeRequestComments = await this.gitlab.MergeRequestNotes.all(
      this.project_id as number, this.iid as number,
    ) as Comment[]
    if (!Array.isArray(currentMergeRequestComments)) {
      throw new Error(
        'Could not find any comments at all. Are you sure you have your request correctly setup?',
      )
    }

    return currentMergeRequestComments
  }

  private async getUser() {
    if (this.user) {
      return this.user;
    }

    const user = await this.gitlab.Users.current({}) as User

    if (user && user.id) {
      this.user = user
    } else {
      throw new Error('Could not retrieve Gitlab user, your token is probably wrong');
    }

    return this.user
  }

  private async createComment(body: string) {
    return this.gitlab.MergeRequestNotes.create(
      this.project_id as number, this.iid as number, body,
    );
  }

  private async createKeyedComment(body: string) {
    this.debug('Creating new Keyed Comment');

    return this.createComment(
      `${this.header}${this.delim}${body}\n${this.delim}${this.footer}`
    );
  }
}
