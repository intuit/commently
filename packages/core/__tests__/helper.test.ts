import { buildKeyedComment } from "../src/helpers";
import { Gitlab, Comment } from "../src/core";

describe('buildKeyedComment', () => {
  let fauxThis : Gitlab
  let fauxComment: Comment
  beforeEach(() => {
    fauxThis = {
      header: 'header',
      footer: 'footer',
      key: 'key',
      delim: 'delim',
      useHistory: true,
      debug: jest.fn()
    } as unknown as Gitlab
    fauxComment = {
      id: 'comment-id',
      body: 'body'
    } as unknown as Comment
  })
  test('creates a comment with history', () => {
    const {id, body} = buildKeyedComment.bind(fauxThis)(fauxComment, 'new body', true)
    expect(id).toBe(fauxComment.id)
    expect(body).toMatchSnapshot()
    expect(fauxThis.debug).toBeCalledTimes(1)
  })
  test('creates a comment without history', () => {
    const {id, body} = buildKeyedComment.bind(fauxThis)(fauxComment, 'new body', false)
    expect(id).toBe(fauxComment.id)
    expect(body).toMatchSnapshot()
    expect(fauxThis.debug).toBeCalledTimes(1)
  })
})