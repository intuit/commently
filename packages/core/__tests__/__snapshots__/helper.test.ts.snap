// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildKeyedComment creates a comment with history 1`] = `
"bodydelimnew body
delim<details><summary>History</summary><div>
<!-- key-history -->
<hr><p></p><!-- key-history -->
</div></details>
</br>"
`;

exports[`buildKeyedComment creates a comment without history 1`] = `
"headerdelimnew body
delimfooter"
`;
